executor/studentapi.pyx

```diff
@@ -0,0 +1,272 @@
+# cython: nonecheck=True
+
+from cython cimport view
+from cpython.mem cimport PyMem_Malloc, PyMem_Free
+
+import threading
+import sys
+import builtins
+# import numpy as np
+
+"""Student API written in Cython. """
+
+# Initializing logger to be used throughout API
+logger_init(STUDENTAPI)
+log_runtime(DEBUG, "Student API intialized")
+
+MAX_THREADS = 8
+
+## Tools used for logging
+
+def _print(*values, sep=' ', end='\n', file=None, flush=None, level=INFO):
+    """Helper print function that redirects message for stdout instead into the Runtime logger. Will print at the INFO level.
+
+    Args:
+        values: iterable of values to print
+        sep: string used as the seperator between each value. Default is ' '
+    """
+    string = sep.join(map(str, values)) + end
+    if file == sys.stdout:
+        level = INFO
+    elif file == sys.stderr:
+        level = ERROR
+    log_runtime(level, string.encode('utf-8'))
+
+class OutputRedirect:
+    def __init__(self, level):
+        self.level = level
+    def write(self, text):
+        log_runtime(self.level, text.encode('utf-8'))
+
+builtins.print = _print
+sys.stderr = OutputRedirect(PYTHON)
+
+
+## Test functions for SHM
+
+def _shm_init():
+    """ONLY USED FOR TESTING. NOT USED IN PRODUCTION"""
+    shm_init(STUDENTAPI)
+    shm_aux_init(STUDENTAPI)
+
+def _shm_stop():
+    """ONLY USED FOR TESTING. NOT USED IN PRODUCTION"""
+    shm_stop(STUDENTAPI)
+    shm_aux_stop(STUDENTAPI)
+
+
+## API Objects
+
+class CodeExecutionError(Exception):
+    """
+    An exception that occurred while attempting to execute student code
+
+    `CodeExecutionError` accepts arbitrary data that can be examined or written into logs.
+
+    Example:
+        >>> err = CodeExecutionError('Error', dev_ids=[2, 3])
+        >>> err
+        CodeExecutionError('Error', dev_ids=[2, 3])
+
+    """
+
+    def __init__(self, message='', **context):
+        super().__init__(message)
+        self.context = context
+
+    def __repr__(self):
+        cls_name, (msg, *_) = self.__class__.__name__, self.args
+        kwargs = ', '.join(f'{name}={repr(value)}' for name, value in self.context.items())
+        return f'{cls_name}({repr(msg)} {kwargs})'
+
+
+cdef class Gamepad:
+    """
+    The API for accessing gamepads.
+
+    Attributes:
+        mode: The execution state of the robot.
+    """
+    cdef public str mode
+
+    def __cinit__(self, mode):
+        """Initializes the mode of the robot. Also initializes the auxiliary SHM. """
+        self.mode = mode
+        # shm_aux_init(STUDENTAPI)
+        # log_runtime(DEBUG, "Aux SHM initialized")
+
+    def __dealloc__(self):
+        """Once process is finished and object is deallocated, close the mapping to the auxiliary SHM."""
+        # shm_aux_stop(STUDENTAPI)
+        # log_runtime(DEBUG, "Aux SHM stopped")
```

Comment:
So we straightup don't need this any more? Do we still need this function?

Reply:
No